src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/cost/RouteCostCalculator.java

```diff
@@ -0,0 +1,93 @@
+package org.opentripplanner.routing.algorithm.raptoradapter.transit.cost;
+
+import java.util.Set;
+import java.util.function.DoubleFunction;
+import javax.annotation.Nonnull;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.raptor.api.transit.CostCalculator;
+import org.opentripplanner.transit.raptor.api.transit.RaptorTransfer;
+import org.opentripplanner.transit.raptor.api.transit.RaptorTransferConstraint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RouteCostCalculator<T extends DefaultTripSchedule> implements CostCalculator<T> {
+
+  public static final double DEFAULT_ROUTE_RELUCTANCE = 1.0;
+  public static final double UNPREFERRED_ROUTE_RELUCTANCE = 2.0;
+
+  private final CostCalculator delegate;
+  private final Set<FeedScopedId> unpreferredRoutes;
+  private final DoubleFunction<Double> unpreferredRouteCost;
+
+  public RouteCostCalculator(
+    @Nonnull CostCalculator<T> delegate,
+    @Nonnull Set<FeedScopedId> routePenalties,
+    @Nonnull DoubleFunction<Double> unpreferredRouteCost
+  ) {
+    this.unpreferredRoutes = routePenalties;
+    this.delegate = delegate;
+    this.unpreferredRouteCost = unpreferredRouteCost;
+  }
+
+  @Override
+  public int boardingCost(
+    boolean firstBoarding,
+    int prevArrivalTime,
+    int boardStop,
+    int boardTime,
+    T trip,
+    RaptorTransferConstraint transferConstraints
+  ) {
+    return delegate.boardingCost(
+      firstBoarding,
+      prevArrivalTime,
+      boardStop,
+      boardTime,
+      trip,
+      transferConstraints
+    );
+  }
+
+  @Override
+  public int onTripRelativeRidingCost(int boardTime, T tripScheduledBoarded) {
+    return delegate.onTripRelativeRidingCost(boardTime, tripScheduledBoarded);
+  }
+
+  @Override
+  public int transitArrivalCost(
+    int boardCost,
+    int alightSlack,
+    int transitTime,
+    T trip,
+    int toStop
+  ) {
+    int unpreferCost = 0;
+    if (unpreferredRoutes.contains(trip.routeId())) {
+      // calculate cost with linear function: fixed + reluctance * transitTime
+      unpreferCost += RaptorCostConverter.toRaptorCost(unpreferredRouteCost.apply(transitTime));
```

Comment:
Currently no preferred mechanism of using _constant_ and _coefficient_ values separately exists. Applying _constant_ part in `boardingCost(...)` would presumably lead better route search performance.